code/modules/projectiles/updated_projectiles/gun_helpers.dm

```diff
@@ -567,48 +539,151 @@ should be alright.
 
 	usr.visible_message("<span class='notice'>[usr] strips [A] from [src].</span>",
 	"<span class='notice'>You strip [A] from [src].</span>", null, 4)
-	A.Detach(src)
+	A.Detach(src, usr)
 
 	playsound(src, 'sound/machines/click.ogg', 15, 1, 4)
 	update_attachables()
 
-/obj/item/weapon/gun/verb/toggle_burst()
+
+/obj/item/weapon/gun/ui_action_click(mob/user, datum/action/item_action/action)
+	if(flags_gun_features & GUN_BURST_FIRING)
+		return
+	var/datum/action/item_action/firemode/firemode_action = action
+	if(!istype(firemode_action))
+		return ..()
+	if(firemode_action.action_firemode == gun_firemode)
+		return
+	gun_firemode = firemode_action.action_firemode
+	playsound(user, 'sound/machines/click.ogg', 15, 1)
+	to_chat(user, "<span class='notice'>[icon2html(src, user)] You switch to <b>[gun_firemode]</b>.</span>")
+	user.update_action_buttons()
+
+
+/obj/item/weapon/gun/verb/toggle_firemode()
 	set category = "Weapons"
-	set name = "Toggle Burst Fire Mode"
-	set desc = "Toggle on or off your weapon burst mode, if it has one. Greatly reduces accuracy."
+	set name = "Toggle Fire Mode"
+	set desc = "Toggle between fire modes, if the gun has more than has one."
 	set src = usr.contents
 
 	var/obj/item/weapon/gun/G = get_active_firearm(usr)
 	if(!G)
 		return
-	src = G
+	G.do_toggle_firemode(usr)
 
-	//Burst of 1 doesn't mean anything. The weapon will only fire once regardless.
-	//Just a good safety to have all weapons that can equip a scope with 1 burst_amount.
-	if(burst_amount < 2)
-		to_chat(usr, "<span class='warning'>This weapon does not have a burst fire mode!</span>")
-		return
 
+/obj/item/weapon/gun/proc/do_toggle_firemode(mob/user, new_firemode)
 	if(flags_gun_features & GUN_BURST_FIRING)//can't toggle mid burst
 		return
 
-	playsound(usr, 'sound/machines/click.ogg', 15, 1)
-	if(flags_gun_features & GUN_HAS_FULL_AUTO)
-		if(flags_gun_features & GUN_BURST_ON)
-			if(flags_gun_features & GUN_FULL_AUTO_ON)
-				flags_gun_features &= ~GUN_FULL_AUTO_ON
-				flags_gun_features &= ~GUN_BURST_ON
-				to_chat(usr, "<span class='notice'>[icon2html(src, usr)] You set [src] to single fire mode.</span>")
-			else
-				flags_gun_features|= GUN_FULL_AUTO_ON
-				to_chat(usr, "<span class='notice'>[icon2html(src, usr)] You set [src] to full auto mode.</span>")
-		else
-			flags_gun_features |= GUN_BURST_ON
-			to_chat(usr, "<span class='notice'>[icon2html(src, usr)] You set [src] to burst fire mode.</span>")
+	switch(length(gun_firemode_list))
+		if(0)
+			CRASH("[src] called do_toggle_firemode() with an empty gun_firemode_list")
```

Comment:
This will happen if `gun_firemode_list` is empty or null, which shouldn't happen and should be caught early.
Since this is a `switch` proc, it's basically a free check.

Reply:
Same as the other.

Reply:
Shouldn't happen != can happen, you didn't answer the question.